src/route/page.rs

```diff
@@ -18,37 +18,134 @@
+                title: "SCP-XXXX",
+                alt_title: Some("The Monster"),
+                header: None,
+                subheader: None,
+                rating: 1000,
+                tags: vec![
+                    "scp",
+                    "keter",
+                    "intangible",
+                    "k-class-scenario",
+                    "ontokinetic",
+                ],
+            };
+
+            let mut output = match renderer.transform(&mut contents, page_info, &remote_handle) {
+                Ok(o) => o,
+                Err(e) => return Ok(HttpResponse::InternalServerError().json(format!("{:?}", e))),
+            };
+
+            // TODO: use jinja/other templater to put this text into a better template
+            let mut buffer = String::from("<html><head>");
+
+            for meta in &output.meta {
+                meta.render(&mut buffer);
+            }
+
+            buffer.push_str("<style>");
+            buffer.push_str(&output.style);
+            buffer.push_str("</style></head><body>");
+            buffer.push_str(&output.html);
+            buffer.push_str("</body></html>\n");
+
+            Ok(HttpResponse::Ok().json(buffer))
+        }
+        Ok(None) => Err(()),
+        Err(e) => {
+            warn!("Failed to retrieve page contents: {}", e);
+
+            Ok(HttpResponse::InternalServerError().json(e))
+        }
+    }
+}
+
+pub async fn get_deepwell_page_wrapped(
+    wiki_id: WikiId,
+    slug: &str,
+    deepwell: web::Data<DeepwellPool>,
+) -> HttpResponse {
+    match get_deepwell_page(wiki_id, slug, &deepwell).await {
+        Ok(o) => o,
+        Err(()) => get_deepwell_page(wiki_id, "_404", &deepwell).await.unwrap(), // todo: make this safer
+    }
+}
+
 /// Route handling for pages, with arguments or not.
-pub async fn page_get(req: HttpRequest) -> HttpResult {
+pub async fn page_get(req: HttpRequest, deepwell: web::Data<DeepwellPool>) -> HttpResult {
     let host = get_host(&req);
     let path = req.uri().path();
 
     info!("GET page {} [{}]", path, host.unwrap_or("none"));
 
-    let _page_req = PageRequest::parse(path);
+    let page_req = PageRequest::parse(path);
 
-    // TODO retrieve page from client
-    Ok(HttpResponse::NotImplemented().finish())
+    Ok(get_deepwell_page_wrapped(*TEMP_WIKI_ID, page_req.slug, deepwell).await)
 }
 
 /// Route for root, which is the same as whatever the `main` page is.
-pub async fn page_main(req: HttpRequest) -> HttpResult {
+pub async fn page_main(req: HttpRequest, deepwell: web::Data<DeepwellPool>) -> HttpResult {
     let host = get_host(&req);
 
     info!("GET / [{}]", host.unwrap_or("none"));
 
-    let _page_req = PageRequest {
+    let page_req = PageRequest {
         slug: "",
         categories: Vec::new(),
         arguments: HashMap::new(),
     };
 
-    // TODO get page request
-    Ok(HttpResponse::NotImplemented().finish())
+    Ok(get_deepwell_page_wrapped(*TEMP_WIKI_ID, page_req.slug, deepwell).await)
```

Comment:
Slug needs to changed from `""`, which is invalid. We need to fetch what the root page is based on wiki settings.